compiler/src/iree/compiler/Dialect/Flow/IR/FlowOps.td

```diff
@@ -125,9 +146,32 @@ def FLOW_DispatchWorkgroupsOp : FLOW_PureOp<"dispatch.workgroups", [
     ValueRange getOperandDynamicDims(unsigned idx) {
       return IREE::Util::findVariadicDynamicDims(idx - getWorkload().size(), getArguments(), getArgumentDims());
     }
+
     ValueRange getResultDynamicDims(unsigned idx) {
       return IREE::Util::findVariadicDynamicDims(idx, getResults(), getResultDims());
     }
+
+    /// Returns the BlockArguments corresponding to the inputs in the type
+    /// signature.
+    Block::BlockArgListType getInputBlockArguments() {
+      return getWorkgroupBody().getArguments().take_front(getArguments().size());
+    }
+
+    /// Returns the BlockArgument corresponding to the idx-th input in the type
+    /// signature.
+    BlockArgument getInputBlockArgument(unsigned idx) {
+      return getWorkgroupBody().getArguments()[idx];
+    }
+
+    /// Returns the BlockArgument corresponding to the idx-th output in the
+    /// type signature. If the output is tied to an input, the returned value
+    /// is also an input BlockArgument.
+    BlockArgument getOutputBlockArgument(unsigned idx);
+
+    /// Returns the BlockArguments corresponding to the outputs in the type
+    /// signature. In case an output is tied to an input, the return list
+    /// overlaps with `getInputBlockArguments`.
+    SmallVector<BlockArgument> getOutputBlockArguments();
```

Comment:
`getInputBlockArguments()` returns `BlockArgListType`. Is it difficult to make this function return the same type?

Reply:
`BlockArgListType` is a wrapper around the underlying data structure (Its like an `ArrayRef` and actually might be typedef-ed to it). So unless it is contiguous in the underlying data structure you cant return that. I think for `getOutput...` its not guaranteed for the arguments to be contiguous. It is for `getInput...`

Reply:
I believe the arguments are laid consecutively, so we should be able to create an array reference from an index to the end with begin()+index and end()?

Reply:
the values for the list to be returned are not contiguous in the original list of arguments.

Reply:
Is there an assumption of how the arguments are laid? I thought that `arguments` is a concat of input and output arguments.

Reply:
But some are tied arguments. Some input arguments are marked as in-out, so the same argument represents both input and output.

Reply:
I see. My assumption was that even that case we have a separate BlockArgument for each. For example, %0, %1, %0, %2, where %0 and %1 are input and %0 and %2 are output. Thanks for the info.